abdiff/core/run_ab_transforms.py

```diff
@@ -78,72 +75,101 @@ def run_ab_transforms(
     """
     start_time = perf_counter()
 
+    # initialize environment
     if not docker_client:
         docker_client = docker.from_env()
-
     transformed_directory_a, transformed_directory_b = create_subdirectories(
         base_directory=run_directory, subdirectories=["transformed/a", "transformed/b"]
     )
     logger.info(
         "Transformed directories created: "
         f"{[transformed_directory_a, transformed_directory_b]}"
     )
-
     run_configs = [
         (image_tag_a, transformed_directory_a),
         (image_tag_b, transformed_directory_b),
     ]
 
-    containers = []
-    for input_file in input_files:
-        source, output_file = parse_transform_details_from_extract_filename(input_file)
-        for docker_image, transformed_directory in run_configs:
-            container = run_docker_container(
-                docker_image,
-                transformed_directory,
-                source,
-                input_file,
-                output_file,
-                docker_client,
-            )
-            containers.append(container)
-
-    failed_containers = wait_for_containers(containers, timeout)
-    logger.info(f"All {len(containers)} containers have exited.")
+    # run containers and collect results
+    futures = run_all_docker_containers(docker_client, input_files, run_configs)
+    containers, exceptions = collect_container_results(futures)
+    logger.info(
+        f"Successful containers: {len(containers)}, failed containers: {len(exceptions)}"
+    )
 
+    # process results
     log_file = aggregate_logs(run_directory, containers)
     logger.info(f"Log file created: {log_file}")
-
-    # errors from failed containers are accessible via aggregated logs
-    if failed_containers:
-        raise DockerContainerRunFailedError(failed_containers)
-
+    if exceptions:
+        raise RuntimeError(  # noqa: TRY003
+            f"{len(exceptions)} / {len(containers) + len(exceptions)} containers failed "
+            "to complete successfully."
+        )
     ab_transformed_file_lists = get_transformed_files(run_directory)
-
     validate_output(ab_transformed_file_lists, len(input_files))
 
+    # write and return results
     run_data = {
         "input_files": input_files,
         "transformed_files": ab_transformed_file_lists,
     }
     update_or_create_run_json(run_directory, run_data)
-
     elapsed_time = perf_counter() - start_time
     logger.info(
         "Total time to complete process: %s", str(timedelta(seconds=elapsed_time))
     )
     return ab_transformed_file_lists
 
 
+def run_all_docker_containers(
+    docker_client: docker.client.DockerClient,
+    input_files: list[str],
+    run_configs: list[tuple],
+) -> list[Future]:
+    """Invoke Docker containers to run in parallel via threads.
+
+    By default, when ThreadPoolExecutor is invoked via a context manager it will wait for
+    all tasks to complete before exiting the context manager.  While each container is run
```

Comment:
This is pretty critical to understanding why this function, `run_all_docker_containers()` is blocking until _all_ containers complete.  

By default, running `ThreadPoolExecutor` via a context manager will wait for all tasks submitted to the `executor` to complete before returning.  You receieve `Future`'s as the result of each thread, so you don't know what happened, but you know the thread is complete.  This allows moving into `collect_container_results()` (which expects a list of `Future`'s) with confidence that all threads are complete.

Reply:
This is all so very timely with our reading!

Reply:
I'll admit I did our reading late... but no kidding!